### ISSUE TITLE
Saving a project forces mixer/automation/* windows to background

### ISSUE THREAD

**Issue by Wallacoloo** [LMMS/lmms#2087]
This only occurs when you have one of the windows in LMMS' maximized, as shown in the gif below (most commonly, people do this with the song editor, but you can do it with the piano roll and automation editors too).

In this case, if I have any other window, like the Mixer, in the foreground and I hit Ctrl+S (or save manually via the file menu), such windows will be sent to the background. It can be an annoyance.

![lmms-ctrl-s](https://cloud.githubusercontent.com/assets/1210751/8072107/efa40494-0f00-11e5-94c5-3641cce82a14.gif)

It is interesting to note that this still occurs if the automation editor or piano roll is maximized instead of the song editor, but there is not issue if no window is maximized.

**Comment by Wallacoloo** [comment]
What's more, this seems to also happen whenever LMMS autosaves - I left the window idle with the mixer on top and every minute on the dot the mixer would be sent to the background (and I would recall it to the foreground manually). Again, this only happened when I had another window inside LMMS maximized.

FURTHERMORE, if another application has focus during the autosave, LMMS forces itself to the foreground (I'm using XFCE as my desktop environment), but again, only if one of the LMMS sub-windows is maximized. This has been plaguing me for weeks, so I'm glad I now know a workaround and have narrowed down the cause.

**Comment by softrabbit** [comment]
I think the first spot where this happens is here:
https://github.com/LMMS/lmms/blob/master/src/core/Song.cpp#L1076

Something in those ```saveState()``` calls does it. Tried commenting out the call to ```gui->pianoRoll()->saveState()``` and LMMS stopped behaving badly when the piano roll was maximised. Will drill deeper later, unless someone else beats me to it.

**Comment by musikBear** [comment]
fyi: not reproduceable on 1.1.3 win 32
(I strongly favor a better behavior for 'smaller windows', in order to bring them forward, and have them staying there. Imo that could be **default behavior**, but a small 'tag-icon' in the window tittle-bar, could work. It is to cumbersome to use the context-menu for every window that opens (imo:) )

**Comment by softrabbit** [comment]
Maybe every window should store its geometry every time it gets a resize or move event and isn't maximized or minimized. And then save that without needing the maximization trickery.

**Comment by Wallacoloo** [comment]
Ohhh. Interesting. So the purpose of the lines around 696 in MainWindow.cpp is to also save the non-maximized/minimized geometry so that pressing the window's restore button will work the same now as when you load the file at a later time.

Perhaps there's a variant of `showMaximized()` that doesn't call the maximized window to focus? Maybe just [setWindowState](http://doc.qt.io/qt-4.8/qwidget.html#setWindowState)?

**Comment by Wallacoloo** [comment]
The other option is to enumerate the `saveWidgetState()` calls by widget depth and hide()/show() every window so that the order is still retained afterward. You may have to additionally direct focus back to the window that originally had focus (not sure). It's a bit wasteful and hacky though, so I think saving the geometry upon resize events would be preferable to this option.

At the very least, we can call `_w->setAttribute(Qt::WA_ShowWithoutActivating)` before the current call to showNormal/Maximized/Minimized so as not to steal focus from other applications, as according to [this](http://stackoverflow.com/questions/966688/show-window-in-qt-without-stealing-focus) post on StackOverflow.

**Comment by Wallacoloo** [comment]
I did some digging to try to figure out how `showNormal()` figures out how to position the window even though `normalGeometry()` reports `QRect(0,0, -1x-1)`.

The code is very simple (from [here](https://github.com/mirror/qt/blob/3593380d9d88d0d73eae9ff87a02ae2d975333ea/src/gui/kernel/qwidget.cpp#L3184)), basically amounting to this:
```c++
void QWidget::showNormal()
{
    ensurePolished();
    setWindowState(windowState() & ~(Qt::WindowMinimized
                                     | Qt::WindowMaximized
                                     | Qt::WindowFullScreen));
    show();
}
```

So it looks like the magic happens somewhere inside the `show()` function.

By the way, the normalGeometry() issue seems to be a relatively obscure bug, only affecting X11 users when the window is maximized by the user and not by a Qt call ([bug report](https://bugreports.qt.io/browse/QTBUG-256)). A user in that bug report suggests doing what amounts to your suggestion of storing the size manually.

**Comment by Wallacoloo** [comment]
I did a bit of toying around with a solution today. I'm not intimately familiar with Qt, but it turns out that all of the LMMS subwindows are created by calls to `gui->mainWindow()->workspace()->addSubWindow(x)`. This corresponds to [QMidArea::addSubWindow](http://doc.qt.io/qt-4.8/qmdiarea.html#addSubWindow), which accepts either a QWidget or a QMdiSubWindow object. In the former case, it will simply instantiate a QMdiSubWindow for the widget. All of our calls currently fall into the former case (passing a QWidget).

So the easiest/best way to fix this is likely to add a new class, `SubWindow` or similar, that inherits from QMdiSubWindow and handles all resize/move events to keep track of the normal geometry. And then abstract the `gui->mainWindow()->workspace()->addSubWindow(x)` call to just `gui->mainWindow()->addSubWindow(QWidget *)`, and wrap the widget in our custom subwindow before forwarding it to the MdiArea addSubWindow function.

Lastly, modify the logic inside of MainWindow::saveWidgetState. It currently has a block that looks like this: 
```c++
if( _w->widget()->parentWidget() != NULL &&
	_w->widget()->parentWidget()->inherits( "QMdiSubWindow" ) )
{
	_w = _w->parentWidget();
}
```

It should only take some minor modifications to make it also use some custom accessors when we inherit from "SubWindow" in place of Qt's x(), y(), etc.

I should hopefully have something ready within a day or two.

**Comment by Wallacoloo** [comment]
Proposed fix via PR #2103.

**Comment by Wallacoloo** [comment]
Closed via #2103.
